### PR TITLE
Feature/graphql

### PR DESCRIPTION
**Pull request by hirsch88**

### CODE REVIEW

File: src/api/mutations/CreatePetMutation.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+@Mutation()
+export class CreatePetMutation extends AbstractGraphQLMutation<GraphQLContext<any, any>, Pet, CreatePetMutationArguments> implements GraphQLFieldConfig {
+    public type = PetType;
+    public args = {
+        name: { type: new GraphQLNonNull(GraphQLString) },
+        age: { type: new GraphQLNonNull(GraphQLInt) },
+    };
+
+    public async run(root: any, args: CreatePetMutationArguments, context: GraphQLContext<any, any>): Promise<Pet> {
+        const petService = context.container.get<PetService>(PetService);
```
Comment (dweber019) [inline src/api/mutations/CreatePetMutation.ts]: Would be nice to have this over DI, but think for the next step.

File: src/lib/graphql/dataloader.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    // same length as the ids. See: https://github.com/facebook/dataloader#batch-function
+    const orderedResult: T[] = [];
+    for (const id of ids) {
+        const item = result.find(t => t[key] === id);
+        if (item) {
+            orderedResult.push(item);
+        } else {
+            /* tslint:disable */
+            // @ts-ignore
+            orderedResult.push(null);
```
Comment (dweber019) [inline src/lib/graphql/dataloader.ts]: Could we use `undefined as any as T`?

File: src/lib/graphql/importClassesFromDirectories.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            exported.forEach((i: any) => loadFileClasses(i, allLoaded));
+        } else if (exported instanceof Object || typeof exported === 'object') {
+            Object.keys(exported).forEach(key => loadFileClasses(exported[key], allLoaded));
+        }
+
+        return allLoaded;
+    };
+
+    const allFiles = directories.reduce((allDirs, dir) => {
+        return allDirs.concat(require('glob').sync(path.normalize(dir)));
```
Comment (dweber019) [inline src/lib/graphql/importClassesFromDirectories.ts]: It there a reason why we don't import glob at the top?

File: test/integration/PetService.test.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import { Pet } from '../../src/api/models/Pet';
+import { PetService } from './../../src/api/services/PetService';
+import { createDatabaseConnection, synchronizeDatabase, closeDatabase } from './utils/database';
+
+describe('PetService', () => {
+
+    let connection: Connection;
+    beforeAll(async () => connection = await createDatabaseConnection());
+    beforeEach(() => synchronizeDatabase(connection));
+    afterAll(() => closeDatabase(connection));
```
Comment (dweber019) [inline test/integration/PetService.test.ts]: Even though the `closeDatabase` function didn't exist before my review commits it has not thrown a error in TS.
Based on this fact any TS errors in integration tests, which don't fail the app hard, will not be detected by the developer.

File: test/integration/utils/database.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        type: env.db.type as any, // See createConnection options for valid types
+        database: env.db.database,
+        logging: env.db.logging,
+        entities: env.app.dirs.entities,
+    });
+    return connection;
+};
+
+export const synchronizeDatabase = (connection: Connection) => {
+    return connection.synchronize(true);
```
Comment (dweber019) [inline test/integration/utils/database.ts]: We should use migrations as sync is not aligned with migrations and migrations is the truth representation of the schema.

File: src/lib/graphql/dataloader.ts (hunk 2 above)
Comment (hirsch88) [reply src/lib/graphql/dataloader.ts]: must be null is given from the lib

**Pull request merged**